alibi/utils/missing_optional_dependency.py

```diff
@@ -0,0 +1,140 @@
+    + "the correct optional dependencies, run `pip install alibi[$missing_dependency]` "
+    + "from the command line. For more information, check the Installation documentation "
+    + "at https://docs.seldon.io/projects/alibi/en/latest/overview/getting_started.html."
+))
+
+
+class MissingDependency(type):
+    """Metaclass for Missing Dependency classes
+
+    Replaces any object that requires unmet optional dependencies. Attribute access or attempting to initialize
+    classes derived from this metaclass will raise an error.
+    """
+    err: Union[ModuleNotFoundError, ImportError]
+    missing_dependency: str = 'all'
+
+    def __new__(mcs, name, bases, attrs):
+        """ Metaclass for MissingDependency classes
+
+        Parameters
+        ----------
+        name:
+            Name of the class to be created
+        bases:
+            Base classes of the class to be created
+        attrs:
+            Attributes of the class to be created, should contain an `err` attribute that will be used to raise an
+            error when the class is accessed or initialized.
+        """
+        return super(MissingDependency, mcs) \
+            .__new__(mcs, name, bases, attrs)
+
+    @property
+    def err_msg(cls):
+        """Generate error message informing user to install missing dependencies."""
+        return err_msg_template.substitute(
+            name=cls.__name__,
+            missing_dependency=cls.missing_dependency)
+
+    def __getattr__(cls, key):
+        """Raise an error when attributes are accessed."""
+        raise ImportError(cls.err_msg) from cls.err
+
+    def __call__(cls, *args, **kwargs):
+        """Raise an error if initialized."""
+        raise ImportError(cls.err_msg) from cls.err
+
+
+class MissingDependencyRay(MissingDependency):
+    missing_dependency = 'ray'
+
+
+class MissingDependencyTensorFlow(MissingDependency):
+    missing_dependency = 'tensorflow'
+
+
+class MissingDependencyTorch(MissingDependency):
+    missing_dependency = 'torch'
+
+
+class MissingDependencyShap(MissingDependency):
+    missing_dependency = 'shap'
+
+
+ERROR_TYPES = {
+    'ray': MissingDependencyRay,
+    'tensorflow': MissingDependencyTensorFlow,
+    'torch': MissingDependencyTorch,
+    'shap': MissingDependencyShap,
+    'numba': MissingDependencyShap,
+}
+
+
+def import_optional(module_name: str, names: Optional[List[str]] = None):
+    """Import a module that depends on optional dependencies
+
+    Note: This function is used to import modules that depend on optional dependencies. Because it mirrors the python
+    import functionality its return type has to be `Any`. Using objects imported with this function can lead to
+    misspecification of types as `Any` when the developer intended to be more restrictive.
+
+    params:
+    ______
+        module:
+            The module to import
+        names:
+            The names to import from the module. If None, all names are imported.
+
+    returns:
+    _______
```

Comment:
Docstring not quite adhering to `numpy` style! (colons not necessary and replace underscores with `=`).

Reply:
Do you mean `-` not `=`?